### PR TITLE
[enhancement] add sklearnex version of ```validate_data```, ```_check_sample_weight```

### PR DESCRIPTION
**Pull request by icfaust**
## Description

This is another interim PR towards introducing the new onedal finiteness checker into the sklearnex estimator workflows. This is not yet introduced into any of the estimators, and so performance benchmarks are not necessary.  This PR focuses on making sure that input and outputs of ```validate_data``` and ```_check_sample_weight``` are respected for sycl_usm_ndarray types and that the new finite checker is properly called and yields results in a range of scenarios.  This is also done to minimize the review burden, as changing all the estimators is a large change.

The new process for all estimators will be as follows:
- All estimators will call ```validate_data``` and ```_check_sample_weight``` once in sklearnex in ```_onedal_*``` methods called by device_offload's dispatch
- All estimators will call assert_all_finite no where else but in ```validate_data``` or ```_check_sample_weight``` unless an operation before the oneDAL backend can yield a inf/NaN (this is a strict condition, and is expected to be extremely uncommon/ hard to allow)
- Calls to check_array anywhere in the ```onedal``` or ```sklearnex``` folders must have assert_all_finite checks turned off.

A follow up PR will create a design test for this, and will introduce the new validate_data in one estimator. Other estimators will occur in individual PRs due to the depth of the changes.

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

**Comment by icfaust** [comment]
Please manually check private CI, failure is a 503 http error (unrelated to the PR).

File: onedal/utils/tests/test_validation.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    [
+        [16, 2048],
+        [
+            2**16 + 3,
+        ],
+        [1000, 1000],
+        [
+            3,
+        ],
+    ],
```
Comment (ahuber21) [inline onedal/utils/tests/test_validation.py:43]:
```suggestion
    "shape",
    [
        [16, 2048],
        [
            2**16 + 3
        ],
        [1000, 1000],
        [
            3
        ],
    ],
```

(removing the trailing commas in the single-element lists will trigger a reformat to a single line)

File: onedal/utils/validation.py [hunk 2]
```diff
@@ -438,3 +443,31 @@ def _is_csr(x):
     return isinstance(x, sp.csr_matrix) or (
         hasattr(sp, "csr_array") and isinstance(x, sp.csr_array)
     )
+
+
+def _assert_all_finite(X, allow_nan=False, input_name=""):
```
Comment (ahuber21) [inline onedal/utils/validation.py:448]: Is a version check required here that sets `_assert_all_finite()` to the `daal4py` version for older oneDAL releases?

File: sklearnex/utils/tests/test_validation.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+)
+
+
+@pytest.mark.parametrize("dtype", [np.float32, np.float64])
+@pytest.mark.parametrize(
+    "shape",
+    [
+        [16, 2048],
+        [
+            2**16 + 3,
```
Comment (ahuber21) [inline sklearnex/utils/tests/test_validation.py:46]:
```suggestion
            2**16 + 3
```

File: sklearnex/utils/tests/test_validation.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    assert type(X_array) == type(X)
+
+
+@pytest.mark.parametrize("dtype", [np.float32, np.float64])
+@pytest.mark.parametrize(
+    "shape",
+    [
+        [16, 2048],
+        [
+            2**16 + 3,
```
Comment (ahuber21) [inline sklearnex/utils/tests/test_validation.py:67]:
```suggestion
            2**16 + 3
```

**Review by ahuber21 — CHANGES_REQUESTED** [review]

File: onedal/utils/validation.py (hunk 2 above)
Comment (icfaust) [reply onedal/utils/validation.py:448]: Changes made for circumstance (won't come up in any of the CIs though)

File: sklearnex/utils/tests/test_validation.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
+# ==============================================================================
+# Copyright 2024 Intel Corporation
```
Comment (napetrov) [inline sklearnex/utils/tests/test_validation.py:2]: Let's go with UXL copyright for new files

File: sklearnex/utils/tests/test_validation.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
+# ==============================================================================
+# Copyright 2024 UXL Foundation Contributors
```
Comment (icfaust) [inline sklearnex/utils/tests/test_validation.py:2]: @napetrov I found only a similar example with a UXL copyright in the oneapi-spec repo in uxlfoundation, let me know if this isn't sufficient.

File: sklearnex/utils/validation.py [hunk 7]
```diff
@@ -14,4 +14,162 @@
...
+    from sklearn.base import BaseEstimator
+
+    _sklearn_validate_data = BaseEstimator._validate_data
+    _finite_keyword = "force_all_finite"
+
+
+if daal_check_version((2024, "P", 700)):
+    from onedal.utils.validation import _assert_all_finite as _onedal_assert_all_finite
+
+    def _onedal_supported_format(X, xp=None):
```
Comment (ethanglaser) [inline sklearnex/utils/validation.py:42]: xp has default None, then xp.float32 is used below, would this lead to an error if function is called without xp specified? Would it make sense to not have a default arg here?

File: sklearnex/utils/validation.py [hunk 8]
```diff
@@ -14,4 +14,162 @@
...
+        **kwargs,
+    )
+    if ensure_all_finite:
+        # run local finite check
+        allow_nan = ensure_all_finite == "allow-nan"
+        arg = iter(out if isinstance(out, tuple) else (out,))
+        if not isinstance(X, str) or X != "no_validation":
+            assert_all_finite(next(arg), allow_nan=allow_nan, input_name="X")
+        if not (y is None or isinstance(y, str) and y == "no_validation"):
+            assert_all_finite(next(arg), allow_nan=allow_nan, input_name="y")
```
Comment (ethanglaser) [inline sklearnex/utils/validation.py]: This is nice but pretty tough to comprehend - a comment about logic would be useful

File: sklearnex/utils/validation.py [hunk 9]
```diff
@@ -14,4 +14,162 @@
...
+
+else:
+    from sklearn.base import BaseEstimator
+
+    _sklearn_validate_data = BaseEstimator._validate_data
+    _finite_keyword = "force_all_finite"
+
+
+if daal_check_version((2024, "P", 700)):
+    from onedal.utils.validation import _assert_all_finite as _onedal_assert_all_finite
```
Comment (ethanglaser) [inline sklearnex/utils/validation.py:40]: Can we add assert_all_finite to onedal.utils init.py? Feel free to use discretion but from my perspective it makes sense to 1. add to init, 2. make assert_all_finite accessible (instead of _assert_all_finite) or otherwise just delete assert_all_finite if its not being used, but you're deeper in the weeds here so let me know what you think makes sense

Comment (ethanglaser) [reply sklearnex/utils/validation.py:40]: And to go further on this - would it make sense to `from sklearn.utils import assert_all_finite` instead of `from sklearn.utils.validation import _assert_all_finite` ? Seems like this is the intended use, but maybe there is overhead with this option, any thoughts?

File: sklearnex/utils/validation.py [hunk 10]
```diff
@@ -14,4 +14,162 @@
...
+        arg = iter(out if isinstance(out, tuple) else (out,))
+        if not isinstance(X, str) or X != "no_validation":
+            assert_all_finite(next(arg), allow_nan=allow_nan, input_name="X")
+        if not (y is None or isinstance(y, str) and y == "no_validation"):
+            assert_all_finite(next(arg), allow_nan=allow_nan, input_name="y")
+    return out
+
+
+def _check_sample_weight(
+    sample_weight, X, dtype=None, copy=False, only_non_negative=False
```
Comment (ethanglaser) [inline sklearnex/utils/validation.py:152]: Any particular reason `only_non_negative` is used here vs. `ensure_non_negative`?

**Review by ethanglaser — COMMENTED** [review]
Initial review - will check test updates once initial comments are addressed. Overall this is excellent work and from my perspective it is very feasible to get merged by end of week.

**Comment by ethanglaser** [comment]
Any perf results to share? (doesn't even have to be full benchmarks but even a large BasicStatistics single GPU run would indicate progress)

File: sklearnex/utils/validation.py (hunk 7 above)
Comment (icfaust) [reply sklearnex/utils/validation.py]: Changed from kwarg to arg.

File: sklearnex/utils/validation.py (hunk 9 above)
Comment (icfaust) [reply sklearnex/utils/validation.py:40]: The underlying ```_assert_all_finite``` is used as ```assert_all_finite``` is a thin wrapper for accessing sparse data when necessary. It is best only to that check once for sparsity at the initial use of ```assert_all_finite```. The original inclusion was made in onedal in order to verify sparse support, and the testing was converted to ```assert_all_finite``` because of this: https://github.com/uxlfoundation/scikit-learn-intelex/pull/2126#discussion_r1861345654 . We can delete the sparse conversion version from onedal with the expectation that we will have to convert the testing there as well.

File: sklearnex/utils/validation.py (hunk 10 above)
Comment (icfaust) [reply sklearnex/utils/validation.py:152]: will change.

**Comment by icfaust** [comment]
> Any perf results to share? (doesn't even have to be full benchmarks but even a large BasicStatistics single GPU run would indicate progress)

I am purposefully not including this in any estimators at this point to speed the review/merging of the PR: there will be performance benchmarks for #2209 #2207 #2206 #2201 and #2189. Good question.

File: sklearnex/tests/utils/base.py [hunk 11]
```diff
@@ -369,3 +372,41 @@ def _get_processor_info():
...
+        y_table = to_table(y)
+        # The presence of the fitted attributes (ending with a trailing
+        # underscore) is required for the correct check. The cleanup of
+        # the memory will occur at the estimator instance deletion.
+        if sua_iface:
+            self.x_attr_ = from_table(
+                X_table, sua_iface=sua_iface, sycl_queue=X.sycl_queue, xp=xp
+            )
+            self.y_attr_ = from_table(
+                y_table, sua_iface=sua_iface, sycl_queue=X.sycl_queue, xp=xp
```
Comment (Vika-F) [inline sklearnex/tests/utils/base.py:391]: Is it Ok that `y_attr_` goes with X's queue?

Also, what happens if `X` and `y` are from different namespaces and have different `sua_iface`? For example, `X` - from dpnp and `y` - from numpy. What is expected to happen in this case?

```suggestion
                y_table, sua_iface=sua_iface, sycl_queue=y.sycl_queue, xp=xp
```

File: sklearnex/utils/validation.py (hunk 8 above)
Comment (icfaust) [reply sklearnex/utils/validation.py]: added.

File: sklearnex/utils/tests/test_validation.py [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+    with config_context(**dispatch):
+        X_out, y_out = validate_data(est, X, y)
+        # check sklearn validate_data operations work underneath
+        X_array = validate_data(est, X, reset=False)
+
+    if dispatch:
+        assert type(X) == type(
+            X_array
+        ), f"validate_data converted {type(X)} to {type(X_array)}"
+        assert type(X) == type(X_out), f"from_array converted {type(X)} to {type(X_out)}"
```
Comment (Vika-F) [inline sklearnex/utils/tests/test_validation.py:232]: I guess `y` needs to be checked here and in 'else' branch as well.

File: sklearnex/tests/utils/base.py (hunk 11 above)
Comment (icfaust) [reply sklearnex/tests/utils/base.py]: This was a failure in the original implementation, though the backend in sklearnex is very fuzzy in this (not standard). For questions about from_table I would ask @samir-nasibli.

File: sklearnex/utils/validation.py (hunk 9 above)
Comment (ethanglaser) [reply sklearnex/utils/validation.py:40]: Thanks for clarification, let's leave onedal4py as is and continue to use _assert_all_finite in cases we know are not sparse then

File: sklearnex/utils/validation.py (hunk 8 above)
Comment (ethanglaser) [reply sklearnex/utils/validation.py]: thanks. lets remove annoyingly (even if it is annoying :) )

**Comment by ethanglaser** [comment]
> > Any perf results to share? (doesn't even have to be full benchmarks but even a large BasicStatistics single GPU run would indicate progress)
> 
> I am purposefully not including this in any estimators at this point to speed the review/merging of the PR: there will be performance benchmarks for #2209 #2207 #2206 #2201 and #2189. Good question. GPU performance improvements will occur when array_api support in the dispatch function is included. (so unfortunately not yet, there is some aspects missing to those PRs which must come from #2096)

Are we expecting perf to be on par with what we were seeing from https://github.com/uxlfoundation/scikit-learn-intelex/pull/2153 ?

File: sklearnex/utils/tests/test_validation.py (hunk 12 above)
Comment (icfaust) [reply sklearnex/utils/tests/test_validation.py]: done

File: sklearnex/utils/validation.py [hunk 13]
```diff
@@ -14,4 +14,194 @@
...
+        **kwargs,
+    )
+
+    check_x = not isinstance(X, str) or X != "no_validation"
+    check_y = not (y is None or isinstance(y, str) and y == "no_validation")
+
+    if ensure_all_finite:
+        # run local finite check
+        allow_nan = ensure_all_finite == "allow-nan"
+        # the return object from validate_data can (annoyingly) be a single
```
Comment (ethanglaser) [inline sklearnex/utils/validation.py:124]:
```suggestion
        # the return object from validate_data can be a single

```

**Review by ethanglaser — APPROVED** [review]
